afkak/test/testutil.py

```diff
@@ -41,20 +55,48 @@ def stat(key, value):
         key, value), file=sys.stderr)
 
 
+def first(deferreds):
+    """Get the first result. Cancel the rest.
+
+    :param deferreds:
+        A sequence of `twisted.internet.defer.Deferred` instances.
+
+        Passing a deferred to *first* transfers ownership: the caller must not
+        add callbacks or cancel it. *first* cancels all other deferreds as soon
+        as one fires.
+
+        This sequence must not be mutated by the caller. *first* does not
+        defensively copy.
+
+    :returns: `Deferred` that fires with the result of the first deferred to
+        fire or fail. Canceling this deferred cancels all of the deferreds.
+    """
+    def cancel_all(self):
+        for d in deferreds:
+            d.cancel()
+
+    result_d = Deferred(cancel_all)
+
+    def one_result(result, source):
+        if result_d.called:
+            return
+        result_d.callback(result)
+        for d in deferreds:
+            if d is not source:
+                d.cancel()
+
+    for d in deferreds:
+        d.addBoth(one_result, d)
+    return result_d
+
+
 # This must only be called from the reactor thread (that is, something
 # decorated with @nose.twistedtools.deferred)
 def async_delay(timeout=0.01, clock=None):
     if clock is None:
         from twisted.internet import reactor as clock
 
-    timeout = timeout
-
-    def succeed():
-        d.callback(timeout)
-
-    d = Deferred()
-    clock.callLater(timeout, succeed)
-    return d
+    return task.deferLater(clock, timeout, lambda: None)
```

Comment:
This makes the deferred returned by async_delay cancellable, so it can't pollute the reactor.